library/psa_crypto.c

```diff
@@ -4071,19 +4050,68 @@ static psa_status_t psa_cipher_setup( psa_cipher_operation_t *operation,
 
     /* A context must be freshly initialized before it can be set up. */
     if( operation->alg != 0 )
-    {
         return( PSA_ERROR_BAD_STATE );
+
+    /* The requested algorithm must be one that can be processed by cipher. */
+    if( ! PSA_ALG_IS_CIPHER( alg ) )
+    {
+        memset( operation, 0, sizeof( *operation ) );
```

Comment:
If I had to guess, maybe this was in a different order of operations before, and the memset was to return the context to a state where a new call to setup (with a correct alg) would succeed instead of throwing a bad state?

I agree it seems rather superfluous now. You want me to remove it?

Reply:
While at it why not. @gilles-peskine-arm what do you think about that?

Reply:
This `memset` was written c06e0712 during the initial implementation of the multipart cipher API. At the time, it was necessary because the API was a little different: there was no `init`, and `setup` functions were guaranteed to initialize the object to a valid state even on failure. This was inconvenient because you couldn't set an object to a valid state until you knew what parameters you wanted to pass to `setup`, and this meant you had to be very careful in cleanup code to only clean up objects that had gone through `setup`. So we changed the API to add `init`, which meant that `setup` no longer had to change the operation object on failure. But we forgot to remove this now-superfluous `memset`.